src/hooks.ts

```diff
@@ -14,24 +14,36 @@ export function useTargetLanguages() {
       .filter(([key]) => key.startsWith('lang'))
       .sort(([key1], [key2]) => key1.localeCompare(key2))
       .map(([_, value]) => value)
-      .filter(i => i && i !== 'none' && i !== 'auto')
+      .filter((i) => i && i !== 'none' && i !== 'auto')
     return Array.from(new Set(langs)) as LanguageCode[]
   }, [])
 }
 
+export const useIsMountedRef = () => {
+  const isMountedRef = useRef(false)
+
+  useEffect(() => {
+    isMountedRef.current = true
+    return () => {
+      isMountedRef.current = false
+    }
+  }, [])
+  return isMountedRef
+}
 export function useSystemSelection() {
   const [text, setText] = React.useState('')
   const preferences = usePreferences()
-
-  if (preferences.getSystemSelection) {
-    React.useEffect(() => {
-      getSelectedText()
-        .then((cbText) => {
-          setText((cbText ?? '').trim())
-        })
-        .catch(() => {})
-    }, [])
-  }
+  const isMountedRef = useIsMountedRef()
+  React.useEffect(() => {
+    if (!preferences.getSystemSelection) {
+      return
+    }
+    getSelectedText()
+      .then((cbText) => {
+        if (isMountedRef.current) setText((cbText ?? '').trim())
+      })
+      .catch(() => {})
+  }, [preferences.getSystemSelection])
```

Comment:
IMHO this is also wrong. Here is a way to implement this without `useRef`:

```suggestion
  React.useEffect(() => {
    if (!preferences.getSystemSelection) {
      return
    }
    let isCancelled = false;
    getSelectedText()
      .then((cbText) => {
        if (isCancelled) setText((cbText ?? '').trim())
      })
      .catch(() => {})
      
    return () => { isCancelled = true };
  }, [preferences.getSystemSelection])
```

Reply:
Learned it!

Reply:
@Innei 

This known pattern can also help avoiding the following race condition:

```tsx
interface ExampleComponentProps {
  dataKey: 'data1' | 'data2'
}

const ExampleComponent = ({ dataKey }: ExampleComponentProps) => {
  const [data, setData] = useState(null);
  useEffect(() => {
    let isCancelled = false;

    someAsyncStuff().then(data => {
      if (!isCancelled) {
        setData(data);
      }
    });

    return () => {
      isCancelled = true;
    };
  }, [dataKey]);
}
```

```
│ Request data 1 ────────────────────────────────────────► data1 response (setData(data1)) │
           │ Request data 2 ────► data2 response (setData(data2)) │
```

`useIsMountedRef` can not help with that.

Reply:
Learned a lot!